File: docs/algorithms/code/algorithms/swapPairs.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function swapPairs(head: ListNode | null): ListNode | null {
  if (!head) {
    return null;
  }

  const newHead = new ListNode(0);
  newHead.next = head;

  let curr = newHead.next,
    prev = newHead;

  while (curr) {
    const first = curr;
    const second = curr.next;
    if (first && second) {
      const temp = second.next;
      second.next = first;
      first.next = temp;
      prev.next = second;

      curr = temp;
      prev = first;
    } else {
      curr = null;
    }
  }

  return newHead.next;
}

File: docs/algorithms/code/algorithms/swapPairs.go
// 思路：每次看 curr.Next 有没有值，有的话就交换，没有的话就返回
// 交换：prev -> curr/first -> second -> third 变成 second -> first/prev -> third/curr
// 复杂度分析：时间复杂度 O(n)，空间复杂度 O(1)
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func swapPairs(head *ListNode) *ListNode {
	if head == nil {
		return head
	}
	tempNode := ListNode{ Val: 0, Next: head }

	prev, curr := &tempNode, head

	for curr != nil && curr.Next != nil {
		// prev -> curr/first -> second -> third
		first, second, third := curr, curr.Next, curr.Next.Next

		// prev -> second/curr -> first -> third
		prev.Next = second
		second.Next = first
		first.Next = third

		// second -> first/prev -> third/curr
		prev = first
		curr = third
	}

	return tempNode.Next
}